crypto/cmp/cmp_vfy.c

```diff
@@ -0,0 +1,868 @@
+            || !X509_STORE_CTX_init(csc, trusted_store,
+                                    cert, ctx->untrusted_certs))
+        goto err;
+
+    valid = X509_verify_cert(csc) > 0;
+
+    /* make sure suitable error is queued even if callback did not do */
+    if (!valid && ERR_GET_REASON(ERR_peek_last_error()) != err)
+        CMPerr(0, err);
+
+ err:
+    X509_STORE_CTX_free(csc);
+    return valid;
+}
+
+/* Helper functions for improving certificate verification error diagnostics */
+static void print_cert(BIO *bio, X509 *cert, unsigned long neg_cflags)
+{
+    if (cert != NULL) {
+        unsigned long flags = ASN1_STRFLGS_RFC2253 | ASN1_STRFLGS_ESC_QUOTE |
+            XN_FLAG_SEP_CPLUS_SPC | XN_FLAG_FN_SN;
+
+        BIO_printf(bio, "    certificate\n");
+        X509_print_ex(bio, cert, flags, ~X509_FLAG_NO_SUBJECT);
+        if (X509_check_issued((X509 *)cert, cert) == X509_V_OK) {
+            BIO_printf(bio, "        self-signed\n");
+        } else {
+            BIO_printf(bio, " ");
+            X509_print_ex(bio, cert, flags, ~X509_FLAG_NO_ISSUER);
+        }
+        X509_print_ex(bio, cert, flags,
+                      ~(X509_FLAG_NO_SERIAL | X509_FLAG_NO_VALIDITY));
+        if (X509_cmp_current_time(X509_get0_notBefore(cert)) > 0)
+            BIO_printf(bio, "        not yet valid\n");
+
+        if (X509_cmp_current_time(X509_get0_notAfter(cert)) < 0)
+            BIO_printf(bio, "        no more valid\n");
+
+        X509_print_ex(bio, cert, flags, ~(neg_cflags));
+    } else {
+        BIO_printf(bio, "    (no certificate)\n");
+    }
+}
+
+static void print_certs(BIO *bio, const STACK_OF(X509) *certs)
+{
+    if (certs != NULL && sk_X509_num(certs) > 0) {
+        int i;
+
+        for (i = 0; i < sk_X509_num(certs); i++) {
+            X509 *cert = sk_X509_value(certs, i);
+            if (cert != NULL)
+                print_cert(bio, cert, 0);
+        }
+    } else {
+        BIO_printf(bio, "    (no certificates)\n");
+    }
+}
+
+static void print_store_certs(BIO *bio, X509_STORE *store)
+{
+    if (store != NULL) {
+        STACK_OF(X509) *certs = ossl_cmp_X509_STORE_get1_certs(store);
+        print_certs(bio, certs);
+        sk_X509_pop_free(certs, X509_free);
+    } else {
+        BIO_printf(bio, "    (no certificate store)\n");
+    }
+}
+
+/*
+ * Diagnostic function that may be registered using
+ * X509_STORE_set_verify_cb(), such that it gets called by OpenSSL's
+ * verify_cert() function at the end of a cert verification as an opportunity
+ * to gather and queue information regarding a (failing) cert verification,
+ * and to possibly change the result of the verification (not done here).
+ * The CLI also calls it on error while cert status checking using OCSP stapling
+ * via a callback function set with SSL_CTX_set_tlsext_status_cb().
+ * returns 0 if and only if the cert verification is considered failed.
+ */
+int OSSL_CMP_print_cert_verify_cb(int ok, X509_STORE_CTX *ctx)
```

Comment:
This looks potentially like something that belongs in the apps rather than in the library.

Reply:
Right, it is typically part of the CLI (where we'll used it in `cmp.c`), 
but since it is useful also to other apps (including a more high-level CMP client lib that I've meanwhile developed) I moved it into the cmp lib. Hope it doesn't hurt here.